### PR TITLE
Add extra information to deploy events, supplemented by updates from porter-agent

### PR DESCRIPTION
**Pull request by ferozemohideen**
## Pull request type

Currently, when a porter app is deployed, a deploy event with status SUCCESS is sent to the server, which includes only the image tag that was deployed and no additional metadata.

With these changes, a deploy event with status PROGRESSING is sent on application deploy; that event also has metadata on the services that are deployed in that revision. The porter-agent is now responsible for listening to the deployment status that is associated with each service deploy, and updating the event if the deployment reaches a successful or failed state.

This feature is only enabled for users with updated porter-agent (code changes to enable the updating behavior are tracked here: https://github.com/porter-dev/porter-agent/pull/52) and, currrently, only porter.run emails. Once this feature is ready to go, we can drop that secondary dependency.

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Other (please describe):

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] If it's a backend change, tests for the changes have been added and `go test ./...` runs successfully from the root folder.
- [ ] If it's a frontend change, Prettier has been run
- [ ] Docs have been reviewed and added / updated if needed

## What is the current behavior?

image

## What is the new behavior?

image

### CODE REVIEW

File: api/server/handlers/cluster/detect_agent_installed.go [hunk 1]
```diff
@@ -32,31 +33,29 @@ func NewDetectAgentInstalledHandler(
...
 	if err != nil {
-		c.HandleAPIError(w, r, apierrors.NewErrInternal(err))
+		err = telemetry.Error(ctx, span, err, "failed to get k8s agent")
+		c.HandleAPIError(w, r, apierrors.NewErrPassThroughToClient(err, http.StatusInternalServerError))
 		return
 	}
 
-	depl, err := agent.GetPorterAgent()
-
+	res, err := GetAgentVersionResponse(agent)
```
Comment (jose-fully-ported) [inline api/server/handlers/cluster/detect_agent_installed.go:48]: Should we be handling the error here?

File: api/server/handlers/porter_app/create.go [hunk 2]
```diff
@@ -297,7 +301,14 @@ func (c *CreatePorterAppHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 			return
 		}
 
-		_, err = createPorterAppEvent(ctx, "SUCCESS", porterApp.ID, 1, imageInfo.Tag, c.Repo().PorterAppEvent())
+		// Only create the PROGRESSING event if the cluster's agent is updated, because only the updated agent can update the status
+		// TODO: remove dependence on porter email once we are ready to release this feature
+		if isPorterAgentUpdated(k8sAgent, 3, 1, 6) && strings.HasSuffix(user.Email, "porter.run") {
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/create.go:306]: Can we change:

```
strings.HasSuffix(user.Email, "porter.run")
```

to

```
isPorterEmail(user.Email)
```

So we don't re-implement that logic everywhere?

File: api/server/handlers/porter_app/create.go [hunk 3]
```diff
@@ -203,7 +207,7 @@ func (c *CreatePorterAppHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 		// create the release job chart if it does not exist (only done by front-end currently, where we set overrideRelease=true)
 		if request.OverrideRelease && preDeployJobValues != nil {
 			telemetry.WithAttributes(span, telemetry.AttributeKV{Key: "installing-pre-deploy-job", Value: true})
-			conf, err := createReleaseJobChart(
+			conf, err := createPreDeployJobChart(
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/create.go:210]: Why this change (same question below)?

File: api/server/handlers/porter_app/create.go (hunk 2 above)
Comment (jose-fully-ported) [reply api/server/handlers/porter_app/create.go:306]: We might also want to wrap this if logic in a helper like:

```
areProgressEventsEnabled()
```

Or something. That centralizes the logic and makes it so we can enable it more widely later on without trying to determine if a particular `if` guard should be updated.

I'd create a `features.go` to centralize these sorts of functions :)

File: api/server/handlers/porter_app/create.go [hunk 4]
```diff
@@ -519,7 +541,117 @@ func createPorterAppEvent(ctx context.Context, status string, appID uint, revisi
...
+	if revision <= 1 {
+		return
+	}
+	revisionFloat64 := float64(revision - 1)
+	matchEvent, err := repo.ReadDeployEventByRevision(ctx, appID, revisionFloat64)
+	if err != nil {
+		_ = telemetry.Error(ctx, span, err, "error reading deploy event by revision")
+		return
+	}
+	if matchEvent.ID == uuid.Nil {
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/create.go:605]: I'm not familiar with `uuid.Nil`, why not just compare to `nil`?

File: api/server/handlers/porter_app/create.go [hunk 5]
```diff
@@ -297,7 +301,14 @@ func (c *CreatePorterAppHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 			return
 		}
 
-		_, err = createPorterAppEvent(ctx, "SUCCESS", porterApp.ID, 1, imageInfo.Tag, c.Repo().PorterAppEvent())
+		// Only create the PROGRESSING event if the cluster's agent is updated, because only the updated agent can update the status
+		// TODO: remove dependence on porter email once we are ready to release this feature
+		if isPorterAgentUpdated(k8sAgent, 3, 1, 6) && strings.HasSuffix(user.Email, "porter.run") {
+			serviceDeploymentStatusMap := getServiceDeploymentMetadataFromValues(values, "PROGRESSING")
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/create.go:307]: Non-blocking, but the status `PROGRESSING` and any other similar statuses (like `SUCCESS`) should probably be a `const`.

File: api/server/handlers/porter_app/create.go [hunk 6]
```diff
@@ -606,3 +738,41 @@ func cloneEnvGroup(c *CreatePorterAppHandler, w http.ResponseWriter, r *http.Req
...
+	}
+
+	parsedTag := strings.Split(tag, ".")
+	if len(parsedTag) != 3 {
+		return false
+	}
+
+	parsedMajor, _ := strconv.Atoi(parsedTag[0])
+	parsedMinor, _ := strconv.Atoi(parsedTag[1])
+	parsedPatch, _ := strconv.Atoi(parsedTag[2])
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/create.go:767]: Will the patch version ever have a tag or something? Like `1+beta3`?

File: api/server/handlers/porter_app/parse.go [hunk 7]
```diff
@@ -941,3 +941,49 @@ func addLabelsToService(service *Service, envGroups []string, defaultLabelKey st
...
+						}
+					}
+				}
+
+				if porterHosts, ok := ingressMap["porter_hosts"].([]interface{}); ok && len(porterHosts) == 1 {
+					return porterHosts[0].(string)
+				}
+			}
+		}
+	}
```
Comment (jose-fully-ported) [inline api/server/handlers/porter_app/parse.go:986]: Non-blocking, but nested if statements like this are... well not great to read :D. I wonder if there is a cleaner way to do this other than exiting early.

File: api/server/handlers/cluster/detect_agent_installed.go (hunk 1 above)
Comment (ferozemohideen) [reply api/server/handlers/cluster/detect_agent_installed.go:48]: oh I'm not sure if I added this after you commented but we handle the error right below

File: api/server/handlers/porter_app/create.go (hunk 2 above)
Comment (ferozemohideen) [reply api/server/handlers/porter_app/create.go:306]: good call! made the file and added the logic there

File: api/server/handlers/porter_app/create.go (hunk 3 above)
Comment (ferozemohideen) [reply api/server/handlers/porter_app/create.go:210]: ah well when I initially wrote this code, I made it so that the internal name for a pre-deploy job was a 'release' job, modeled after Heroku's 'release' keyword for similar functionality. I only realized later that the word 'release' is pretty overloaded in the repo, referring to Helm releases - this is my attempt to correct the ambiguity I created :/

File: api/server/handlers/porter_app/create.go (hunk 4 above)
Comment (ferozemohideen) [reply api/server/handlers/porter_app/create.go:605]: I believe this is a gorm intricacy, I get a compile error when I try to compare to `nil` because of mismatched types. I think with the gorm find method we're using, it won't return an error if the event is not found, but rather an empty row with uuid.Nil 🤷

File: api/server/handlers/porter_app/create.go (hunk 5 above)
Comment (ferozemohideen) [reply api/server/handlers/porter_app/create.go]: good point, created a type for the status and made each one a const!

File: api/server/handlers/porter_app/create.go (hunk 6 above)
Comment (ferozemohideen) [reply api/server/handlers/porter_app/create.go]: Hmm currently we only release agent versions that have numerical patch versions, but I see what you're saying here. Ideally we can remove this dependency entirely once we go about upgrading everyone's agent versions (I added a note about monitoring telemetry and removing this once we see 100% adoption of the new version)

File: api/server/handlers/cluster/detect_agent_installed.go (hunk 1 above)
Comment (jose-fully-ported) [reply api/server/handlers/cluster/detect_agent_installed.go:48]: Ah gotcha, the `targetErr` thing confused me.

File: api/server/handlers/cluster/detect_agent_installed.go [hunk 8]
```diff
@@ -32,31 +33,29 @@ func NewDetectAgentInstalledHandler(
...
 	if targetErr := kubernetes.IsNotFoundError; err != nil && errors.Is(err, targetErr) {
```
[the remark this answers is not in the document]
Comment (ferozemohideen) [reply api/server/handlers/cluster/detect_agent_installed.go]: haha yep that's clearer, fixed

File: dashboard/src/components/porter/Link.tsx [hunk 9]
```diff
@@ -50,28 +54,46 @@ const Svg = styled.svg`
...
   height: 1px;
   width: 100%;
-  background: #ffffff;
+  background: ${(props) => props.color ?? "#ffffff"};
 `;
 
-const LinkWrapper = styled.span`
-  position: relative;
+const StyledLink = styled(DynamicLink) <{ hasunderline?: boolean, color?: string }>`
+  color: ${(props) => props.color ?? "#ffffff"};
```
Comment (jose-fully-ported) [inline dashboard/src/components/porter/Link.tsx:67]: React noob here. Is there a way to specify the defaults in the arguments vs inline in code?

File: dashboard/src/components/porter/Link.tsx [hunk 10]
```diff
@@ -50,28 +54,46 @@ const Svg = styled.svg`
...
-const StyledLink = styled(DynamicLink)<{ hasunderline?: boolean }>`
-  color: #ffffff;
+const LinkWrapper = styled.span<{ hoverColor?: string, color?: string }>`
+  position: relative;
   display: inline-flex;
-  font-size: 13px;
-  cursor: pointer;
+  align-items: center;
+  :hover {
+    ${StyledLink} {
```
Comment (jose-fully-ported) [inline dashboard/src/components/porter/Link.tsx:87]: React noob: What does this `${Name}` thing do?

File: dashboard/src/main/home/app-dashboard/expanded-app/activity-feed/events/cards/ServiceStatusDetail.tsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            case "SUCCESS":
+                return "DEPLOYED";
+            case "FAILED":
+                return "FAILED";
+            case "CANCELED":
+                return "CANCELED";
+            default:
+                return "UNKNOWN";
+        }
+    };
```
Comment (jose-fully-ported) [inline dashboard/src/main/home/app-dashboard/expanded-app/activity-feed/events/cards/ServiceStatusDetail.tsx:35]: For future reference, we'll want to ensure these are properly set in the api in the future vs translated here in the frontend.

File: dashboard/src/components/porter/Link.tsx (hunk 9 above)
Comment (ferozemohideen) [reply dashboard/src/components/porter/Link.tsx:67]: good call, specified the default in the component props

File: dashboard/src/components/porter/Link.tsx (hunk 10 above)
Comment (ferozemohideen) [reply dashboard/src/components/porter/Link.tsx:87]: that applies the style to every child styled component. I tried adding the color and background-color changes there without specifying the child component, but for some reason they weren't propagating to the children :/ this was the only way that worked

i am also very much a css noob lol

**Review by jose-fully-ported — APPROVED** [review]
